src/inference/AdvancedSMC.jl

```diff
@@ -387,7 +387,7 @@ function resample_residual(w::AbstractVector{<:Real}, num_particles::Integer)
     M = length(w)
 
     # "Repetition counts" (plus the random part, later on):
-    Ns = floor.(length(w) .* w)
+    Ns = floor.(M .* w)
```

Comment:
Actually now that I look at the implementation I think we might want to change it even more and make it a bit more efficient.
I'd suggest the following (replacing also the lines below):
```suggestion
    M_w = M .* w
    Ns = floor(Int, M_w)
    
    # The "remainder" or "residual" count
    R = sum(Ns)
    
    # The number of particles which will be drawn stochastically:
    M_rdn = num_particles - R

    # The modified weights:
    Ws = @. (M_w - Ns) / M_rdn
```
Currently `Ns` is an array of floating point numbers, which is not needed and makes all computations involving `Ns` such as `sum(Ns)` and `1:Ns[j]` operate on floating point numbers instead of integers. Furthermore, the computation of `M .* w` is repeated three times currently.

Reply:
Maybe even more efficiently would be something like
```suggestion
    residuals = similar(w)
    indices = similar(Int, w)

    # deterministic assignment
    i = 1
    m = length(w)
    @inbounds for j in 1:m
        x = m * w[i]
        floor_x = floor(Int, x)
        for k in 1:floor_x
            indices[i] = j
            i += 1
        end
        residuals[j] = x - floor_x
    end
    
    # sampling from residuals
    residuals ./= sum(residuals)
    rand!(Categorical(residuals), view(indices, i:m))
    
    return indices
end
```

That avoids the creation of all the intermediate arrays, the computation of how many particles are assigned deterministically, and in particular avoids the possibly costly call of `append!` in the end.